ourjapanlife/findadoc-web
# FEAT: Create login page and auth store

Resolves #575
Resolves #449
Part of epic #542

# 🔧 What changed
###general improvements
- ~~Fixed `$t()` and `import .. from '#app'` compiler errors! 🙌 🥳  this was by adding nuxt support to our tsconfig through the extension. Had to do some moving around of the settings to get it to work with what we already had configured.~~ these changes were moved to #560  and #566.
- improved the loading icon to have less padding
- improved the loading UI for results 
- added an error theme color (for validation error messages)

### New logged in experience
- added links to topnav and hamburger menu for when a user is logged in for convenience! 🏖️ 
- added friendly un-authorized messages to the mod dashboard if someone tries to view without permissions
- reworked the moderation dashboard access to just be based on logged in status instead of the previous env flag + route logic 

### created login page! 
- added validation
- made validation only run after typing has happened
- connected to auth store and loading state
- added redirection to moderation panel once logged in
- made a nice logging in loading indicator on the button

### created auth store
- now have a global state for checking the users logged in status
- has login function written (minus actual API call, since it doesn't exist yet)
- set up loading state events
- added roles concept and built out schema

### auth framework
- Added in supertokens logic (initializiation, supertokens ui, thirdparty login)....then ripped it all out and switched to [Auth0](https://auth0.com/). 
  - Why? supertokens didn't have great vue support and I spent several hours getting it to work. I would have gone with auth0 to start but they're expensive. I recently discovered they have NPO discounts, so it's an option again! 
- Added in auth0
  - Created initialization 
  - found a race-condition bug with their intiailization happening after vue lifecycle hooks causing weird errors. built a shared singleton to fix this (as per their recommendation) this fixed it 🐛 🔨 
  - found a nuxt bug where the initialization breaks because nuxt is try to run it at compile/ssr time. 🙄  Fixed the issue using `process.client`. This was hard to track down as the errors were not intuitive.  🐛 🔨 
  - integrated auth0 with api requests
  - Set up third party login
  - Integrated auth0 with login page and `authStore.ts` 

### moderation panel
- added check to redirect to login page if user is not logged in
- added `<Suspense>` which is a nice way to prevent the page from loading until we've confirmed they're logged in

# 🧪 Testing instructions
- try logging in (ask Phil for account) 
- ensure you can use the moderation dashboard. 
- ensure you can't use it without being logged in

# 📸 Screenshots

![image](https://github.com/ourjapanlife/findadoc-web/assets/4411499/38ae4f1d-7bd1-4132-9aff-64ff532f71fc)
